framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeMirror.java

```diff
@@ -1458,8 +1459,10 @@ private void fixupBoundAnnotations() {
             // annotations for all hierarchies, so we don't want to avoid placing bottom on the
             // lower bound for those hierarchies that don't have a qualifier in primaryAnnotations.
             if (!this.getAnnotationsField().isEmpty()) {
+                Set<AnnotationMirror> newAnnos =
```

Comment:
It seems rather confusing that this replacement happens here only for type parameters.
Have you tried replacing all `PolyAlls` in `AnnotatedTypeMirror`, so that we never create an ATM with `PolyAll` in the first place?
Then we would not need the new `PolyAllReplacer` and would handle all types consistently.
At the least, I think adding an explanation to the code where which substitutions happen would be good.

Reply:
`@PolyAll` can only be replace with type system specific poly qualifiers after all other explicit qualifiers have been added to the type. Otherwise, a poly qualifier might be added for a type system that has a different explicit annotation.  So, we can't replace all `@PolyAll` in `AnnotatedTypeMirror`.  In this case, we can because if another annotation is added the poly qualifier is replaced.  (I'll this explanation as a comment.)